lib/realtime/vehicle.ex

```diff
@@ -63,19 +87,136 @@ defmodule Realtime.Vehicle do
     }
   """
   @spec decode(term()) :: t()
-  def decode(%{} = json) do
+  def decode(%{} = json, opts \\ []) do
+    opts = Keyword.merge(@default_opts, opts)
+    stops_on_route_fn = Keyword.get(opts, :stops_on_route_fn)
+    timepoints_on_route_fn = Keyword.get(opts, :timepoints_on_route_fn)
+
+    id = json["id"]
+    label = json["vehicle"]["vehicle"]["label"]
+    timestamp = json["vehicle"]["timestamp"]
+    direction_id = json["vehicle"]["trip"]["direction_id"]
+    route_id = json["vehicle"]["trip"]["route_id"]
+    trip_id = json["vehicle"]["trip"]["trip_id"]
+    current_status = decode_current_status(json["vehicle"]["current_status"])
+    stop_id = json["vehicle"]["stop_id"]
+
+    stops_on_route =
+      try do
+        stops_on_route_fn.(route_id)
+      catch
+        # Handle Gtfs server timeouts gracefully
+        :exit, _ ->
+          []
+      end
+
+    timepoints_on_route =
+      try do
+        timepoints_on_route_fn.(route_id)
+      catch
+        # Handle Gtfs server timeouts gracefully
+        :exit, _ ->
+          []
+      end
+
+    ordered_stops = if direction_id == 0, do: Enum.reverse(stops_on_route), else: stops_on_route
```

Comment:
We can get the stops and timepoints by `trip_id` instead of by `route_id`, which avoids some of the messiness with fetching by route.